pkg/validate/result.go

```diff
@@ -48,131 +51,155 @@ func NewResult() *Result {
-		switch lvl {
-		case logrus.InfoLevel:
-			logger.Info(obj.Message)
-		case logrus.WarnLevel:
-			logger.Warn(obj.Message)
-		case logrus.ErrorLevel:
-			logger.Error(obj.Message)
-		default:
-			return fmt.Errorf("unknown output level %q", obj.Type)
-		}
-	}
-
-	return nil
+// Combine creates a new Result and calls Result.Combine(results).
+func Combine(results ...Result) (r Result, err error) {
+	err = r.Combine(results...)
+	return r, err
 }
 
-// printJSON will print the output in JSON format
-func (o *Result) printJSON() error {
-	prettyJSON, err := json.MarshalIndent(o, "", "    ")
-	if err != nil {
-		return fmt.Errorf("error marshaling JSON output: %v", err)
+// Combine combines results into r, setting r.Passed = false if
+// any Result in results has r.Passed == false.
+func (r *Result) Combine(results ...Result) error {
+	for _, result := range results {
+		r.Outputs = append(r.Outputs, result.Outputs...)
+	}
+	if err := r.prepare(); err != nil {
+		return err
 	}
-	fmt.Printf("%s\n", string(prettyJSON))
 	return nil
 }
 
 // prepare should be used when writing an Result to a non-log writer.
 // it will ensure that the passed boolean will properly set in the case of the setters were not properly used
-func (o *Result) prepare() error {
-	o.Passed = true
-	for i, obj := range o.Outputs {
+func (r *Result) prepare() error {
+	r.Passed = true
+	for i, obj := range r.Outputs {
 		lvl, err := logrus.ParseLevel(obj.Type)
 		if err != nil {
 			return err
 		}
-		if o.Passed && lvl == logrus.ErrorLevel {
-			o.Passed = false
+		if r.Passed && lvl == logrus.ErrorLevel {
+			r.Passed = false
 		}
 		lvlBytes, _ := lvl.MarshalText()
-		o.Outputs[i].Type = string(lvlBytes)
+		r.Outputs[i].Type = string(lvlBytes)
 	}
 	return nil
 }
 
+// Mockable exit function for tests.
+var exit = os.Exit
+
 // PrintWithFormat prints output to w in format, and exits if some object in output
 // is not in a passing state.
-func (o *Result) PrintWithFormat(format string) (err error) {
+func (r *Result) PrintWithFormat(format string) (err error) {
+	return r.printWithFormat(os.Stdout, format)
+}
+
+func (r *Result) printWithFormat(w io.Writer, format string) (err error) {
 	// the prepare will ensure the result data if the setters were not used
-	if err = o.prepare(); err != nil {
-		return fmt.Errorf("error to prepare output: %v", err)
+	if err = r.prepare(); err != nil {
+		return fmt.Errorf("error preparing output: %v", err)
 	}
 
-	printf := o.getPrintFuncFormat(format)
-	if err = printf(o); err == nil && !o.Passed {
-		os.Exit(1) // Exit with error when any Error type was added
+	switch format {
+	case JSONAlpha1Output:
+		err = r.printJSON(w)
+	case TextOutput: // Text
+		// Address all to the Stdout when the type is not JSON
+		entry := logrus.NewEntry(newLoggerTo(w))
+		err = r.printText(entry)
+	default:
+		return fmt.Errorf("invalid result format type: %s", format)
 	}
+	if err == nil && !r.Passed {
+		// Exit when any Error type was added
+		exit(1)
```

Comment:
Is there a reason we hard exit here versus returning an error?

Reply:
@ryantking The only thing I saw was that it was a way to short circuit when a failed test was found. I tried removing it and the output changed. As you can see below, by returning an error we end up also logging an error in addition to the test failure. We could just make it an informative error.

### WITHOUT `exit(1)`

Return an error instead. The error we returned was "test failed".

```
operator-sdk bundle validate --alpha-select-external pkg/validate/testdata/fails.sh ~/dev/gatekeeper-operator/bundle
ERRO[0000] Error: Value : invalid field Pesce
FATA[0000] test failed
```

Error code: 1

---

### With the `exit(1)` 

In this case we keep it as is with the `exit(1)` short circuit.

```
operator-sdk bundle validate --alpha-select-external pkg/validate/testdata/fails.sh ~/dev/gatekeeper-operator/bundle
ERRO[0000] Error: Value : invalid field Pesce
```

Error code: 1

Reply:
I do agree that doing a hard exit seems weird and bad if this gets used some place else as a generic function call.

Reply:
@jmrodri It looks like the root of the problem is that the validator is failing, but not returning an error. I would think `err == nil && !r.Passed` would be impossible. Instead of logging the invalid field, wouldn't we want to bubble it back up and let the caller determine what to do with it?

Reply:
@ryantking I changed the method return signature. It now returns `bool, error`. That way there is no adverse `exit`. And the caller can use the boolean value to determine if there were any test failures.